servers/gateway/handlers/summary.go

```diff
@@ -34,67 +43,206 @@ type PageSummary struct {
+		return
+	}
+
+	resp, err := fetchHTML(pageURL)
+	if err != nil {
+		log.Fatalf("error fetching URL: %v\n", err)
+	}
+
+	PageSummary, err := extractSummary(pageURL, resp)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("error extracting summary: %v", err), http.StatusBadRequest)
+		return
+	}
+
+	defer resp.Close()
+
+	w.Header().Add("Content-Type", "application/json")
+	json.NewEncoder(w).Encode(PageSummary)
 }
 
 //fetchHTML fetches `pageURL` and returns the body stream or an error.
 //Errors are returned if the response status code is an error (>=400),
 //or if the content type indicates the URL is not an HTML page.
 func fetchHTML(pageURL string) (io.ReadCloser, error) {
-	/*TODO: Do an HTTP GET for the page URL. If the response status
-	code is >= 400, return a nil stream and an error. If the response
-	content type does not indicate that the content is a web page, return
-	a nil stream and an error. Otherwise return the response body and
-	no (nil) error.
-
-	To test your implementation of this function, run the TestFetchHTML
-	test in summary_test.go. You can do that directly in Visual Studio Code,
-	or at the command line by running:
-		go test -run TestFetchHTML
-
-	Helpful Links:
-	https://golang.org/pkg/net/http/#Get
-	*/
-	return nil, nil
+	resp, err := http.Get(pageURL)
+
+	if err != nil {
+		log.Fatalf("error fetching URL: %v\n", err)
+	}
+	if resp.StatusCode >= 400 {
+		return nil, fmt.Errorf("response status code was %d", resp.StatusCode)
+	}
+
+	ctype := resp.Header.Get("Content-Type")
+	if !strings.HasPrefix(ctype, "text/html") {
+		return nil, fmt.Errorf("response content type was %s, not text/html", ctype)
+	}
+
+	return resp.Body, err
 }
 
 //extractSummary tokenizes the `htmlStream` and populates a PageSummary
 //struct with the page's summary meta-data.
 func extractSummary(pageURL string, htmlStream io.ReadCloser) (*PageSummary, error) {
-	/*TODO: tokenize the `htmlStream` and extract the page summary meta-data
-	according to the assignment description.
-
-	To test your implementation of this function, run the TestExtractSummary
-	test in summary_test.go. You can do that directly in Visual Studio Code,
-	or at the command line by running:
-		go test -run TestExtractSummary
-
-	Helpful Links:
-	https://drstearns.github.io/tutorials/tokenizing/
-	http://ogp.me/
-	https://developers.facebook.com/docs/reference/opengraph/
-	https://golang.org/pkg/net/url/#URL.ResolveReference
-	*/
-	return nil, nil
+
+	summary := &PageSummary{}
+
+	tokenizer := html.NewTokenizer(htmlStream)
+
+	for {
+		tokenType := tokenizer.Next()
+		if tokenType == html.ErrorToken {
+			err := tokenizer.Err()
+			if err == io.EOF {
+				break
+			}
+			log.Fatalf("error tokenizing HTML: %v", tokenizer.Err())
+		}
+
+		token := tokenizer.Token()
+		if tokenType == html.StartTagToken || tokenType == html.SelfClosingTagToken {
+			if token.Data == "meta" {
+				prop, _ := getAttr(token, "property")
```

Comment:
It would be better to get all the attributes first instead of looping through them all 3 times -0